File: dotnet/source/examples/Features/TopicViews.cs
/**
 * Copyright © 2022 Push Technology Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using System.Threading;
using PushTechnology.ClientInterface.Client.Factories;
using PushTechnology.ClientInterface.Client.Features;
using PushTechnology.ClientInterface.Client.Topics;
using PushTechnology.ClientInterface.Client.Session;
using static System.Console;
using static PushTechnology.ClientInterface.Examples.Runner.Program;
using System.Linq;

namespace PushTechnology.ClientInterface.Example.Features {
    /// <summary>
    /// Client implementation that creates and manages topic views.
    /// </summary>
    public sealed class TopicViews : IExample
    {
        /// <summary>
        /// Runs the topic views example.
        /// </summary>
        /// <param name="cancellationToken">A token used to end the client example.</param>
        /// <param name="args">A single string should be used for the server url.</param>
        public async Task Run(CancellationToken cancellationToken, string[] args)
        {
            var serverUrl = args[0];
            var session = Diffusion.Sessions.Principal("control").Password("password")
                .CertificateValidation((cert, chain, errors) => CertificateValidationResult.ACCEPT)
                .Open(serverUrl);

            // Create a topic
            var topic = "a/path/0";

            try
            {
                await session.TopicControl.AddTopicAsync(topic, TopicType.JSON, cancellationToken);
            }
            catch (Exception ex)
            {
                WriteLine($"Failed to add topic '{topic}' : {ex}.");
                session.Close();
                return;
            }

            try
            {
                var specification = "map ?a/path/ to prefix/<path(0)>";

                ITopicView view1 = await session.TopicViews.CreateTopicViewAsync("View1", specification, CancellationToken.None);
                WriteLine($"Topic view '{view1.Name}' was created.");

                ITopicView view2 = await session.TopicViews.CreateTopicViewAsync("View2", specification, CancellationToken.None);
                WriteLine($"Topic view '{view2.Name}' was created.");
            }
            catch (Exception ex)
            {
                WriteLine($"Topic view could not be created : {ex}.");
                session.Close();
                return;
            }

            try
            {
                var listTask = await session.TopicViews.ListTopicViewsAsync(CancellationToken.None);
                List<ITopicView> views = listTask.ToList();

                WriteLine($"Listing topic views - {views.Count} views were found:");

                foreach (var view in views)
                {
                    WriteLine($"Topic view '{view.Name}'");
                }
            }
            catch (Exception ex)
            {
                WriteLine($"Topic views could not be listed : {ex}.");
            }

            try
            {
                ITopicView view1 = await session.TopicViews.GetTopicViewAsync("View1", CancellationToken.None);
                WriteLine($"Topic view '{view1.Name}' was retrieved with specification '{view1.Specification}' and roles:");

                foreach (var role in view1.Roles)
                {
                    WriteLine($"'{role}'");
                }
            }
            catch (Exception ex)
            {
                WriteLine($"Topic view 'View1' could not be retrieved : {ex}.");
                session.Close();
                return;
            }

            try
            {
                await session.TopicViews.RemoveTopicViewAsync("View1", CancellationToken.None);
                WriteLine($"Topic view 'View1' was removed.");
                await session.TopicViews.RemoveTopicViewAsync("View2", CancellationToken.None);
                WriteLine($"Topic view 'View2' was removed.");
            }
            catch (Exception ex)
            {
                WriteLine($"Topic view could not be removed : {ex}.");
            }

            // Close the session
            session.Close();
        }
    }
}

File: js/examples/topicViews.js
/*******************************************************************************
 * Copyright (C) 2020 - 2023 DiffusionData Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/

const diffusion = require('diffusion');
let session;

async function createTopicView() {
    // First create a topic called 'some/topic'.
    await session.topics.add('some/topic', diffusion.topics.TopicType.STRING);

    // Now add a topic view.
    // This will create a topic 'other/topic' that mirrors the value of 'some/topic'
    await session.topicViews.createTopicView('example-view', 'map ?some/ to other/<path(1)>');
}

async function listTopicViews() {
    // list all topic views
    const views = await session.topicViews.listTopicViews();

    // show the name and specification of each topic view
    console.log('All Topic Views:');
    views.forEach((topicView) => {
        console.log(`${topicView.name}: ${topicView.specification}`);
    });
}

async function getTopicView() {
    // get a topic view
    const topicView = await session.topicViews.getTopicView('example-view');

    // show the name and specification of the topic view
    console.log('Topic View:');
    console.log(`${topicView.name}: ${topicView.specification}`);
}

async function removeTopicViews() {
    // remove a named topic view
    await session.topicViews.removeTopicView('example-view');
}

// Connect to the server. Change these options to suit your own environment.
// Node.js does not accept self-signed certificates by default. If you have
// one of these, set the environment variable NODE_TLS_REJECT_UNAUTHORIZED=0
// before running this example.
diffusion.connect({
    host   : 'diffusion.example.com',
    port   : 443,
    secure : true,
    principal : 'control',
    credentials : 'password'
})
.then(function(sess) {
    session = sess;
})
.then(createTopicView)
.then(listTopicViews)
.then(getTopicView)
.then(removeTopicViews);
